### PR TITLE
Disables usage of parenthesis in unit enum variants.

### LINKED ISSUE #3375: Should `Option::None()` compile?
The following code compiles:
```rust
script;

fn main() {
    let n: Option<u64> = Option::None(); 
}
```
but does not in Rust. The `()` syntax is somewhat justified here because the `None` variant does not require a value and this is equivalent to requiring a unit type. That being said, should we still disallow this, just like Rust?

### LINKED ISSUE #3585: Cannot use turbofish generic syntax
Take the following code:

```rust
script;

pub enum Option<T> {
    None: (),
    Some: T,
}

impl<T> Option<T> {
    pub fn none(self) -> Option<T> {
        Option::<T>::None
    }
}

fn main() {
    Option::Some(5).none();
}
```

Works fine in Rust but gives the following error in Sway:

```
error
  --> /sway/test-case/src/main.sw:10:17
   |
 8 | 
 9 |     pub fn none(self) -> Option<T> {
10 |         Option::<T>::None
   |                 ^^^ generics are not supported here
11 |     }
12 | }
   |
____
```

### PR DESCRIPTION
**Pull request by esdrubal**
With these changes it is no longer possible to do `Option::None()` instead it should be done with `Option::None`.

This changes also fixes enum turbofish without parenthesis so `Option::None::<T>` can be used.

Closes #3585
Closes #3375

### CODE REVIEW

File: sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs [hunk 1]
```diff
@@ -1161,6 +1162,7 @@ impl ty::TyExpression {
             (false, Some(func_decl), None, None) => {
                 warnings.append(&mut function_probe_warnings);
                 errors.append(&mut function_probe_errors);
+                let args = args.unwrap_or_default();
```
Comment (emilyaherbert) [inline sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs:1165]: Is there any case in which this assumption would be incorrect? i.e. is there any code that the user could write that would trigger this to fail silently?

File: sway-error/src/error.rs [hunk 2]
```diff
@@ -416,6 +416,8 @@ pub enum CompileError {
     MoreThanOneEnumInstantiator { span: Span, ty: String },
     #[error("This enum variant represents the unit type, so it should not be instantiated with any value.")]
     UnnecessaryEnumInstantiator { span: Span },
+    #[error("`{ty}` is a unit enum variant, and does not take parentheses to be constructed.")]
```
Comment (emilyaherbert) [inline sway-error/src/error.rs:419]: Can we reword this to:

"The enum variant `{ty}` is of type `unit`, so its constructor does not take arguments or parenthesis. Try removing the `()`."

File: sway-core/src/semantic_analysis/ast_node/expression/typed_expression/enum_instantiation.rs [hunk 3]
```diff
@@ -35,6 +35,19 @@ pub(crate) fn instantiate_enum(
         errors
     );
 
+    if type_engine
```
Comment (emilyaherbert) [inline sway-core/src/semantic_analysis/ast_node/expression/typed_expression/enum_instantiation.rs:38]: Would you add a line comment above this indicating what this block of code is doing?

File: sway-lsp/src/traverse/parsed_tree.rs [hunk 4]
```diff
@@ -624,8 +624,9 @@ impl<'a> ParsedTree<'a> {
                     self.collect_type_arg(type_arg, &token);
                 }
 
-                for exp in args {
-                    self.handle_expression(exp);
+                let args_vec = args.clone().unwrap_or_default();
+                for exp in args_vec {
```
Comment (emilyaherbert) [inline sway-lsp/src/traverse/parsed_tree.rs:628]: To eliminate the `.clone()`, change this block of code to something like:

```rust
args.map(|args_vec| {
  args_vec.iter().for_each(|exp| {
    self.handle_expression(&exp);
  })
});
```

**Review by emilyaherbert — CHANGES_REQUESTED** [review]
Great solution to these two GH issues!

Would you add additional tests to cover these cases:
1. Test when enums are specified with their fully qualified path including the library names both with and without the `()` (`lib_a::inner_lib::my_enum::MyEnum::VariantA()`)
2. Test function calls and method calls that take no arguments both with and without the `()` and both with and without their fully qualified paths including the library names

Thanks!

File: sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs (hunk 1 above)
Comment (esdrubal) [reply sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs]: It was possible to have args set to None and we could have functions called with no parenthesis. Thanks for noticing. Recent changes address this issue, and add a test case as well.

File: sway-error/src/error.rs (hunk 2 above)
Comment (esdrubal) [reply sway-error/src/error.rs]: Applied the suggested change.

File: sway-core/src/semantic_analysis/ast_node/expression/typed_expression/enum_instantiation.rs (hunk 3 above)
Comment (esdrubal) [reply sway-core/src/semantic_analysis/ast_node/expression/typed_expression/enum_instantiation.rs:40]: Added a line comment.

File: sway-lsp/src/traverse/parsed_tree.rs (hunk 4 above)
Comment (esdrubal) [reply sway-lsp/src/traverse/parsed_tree.rs]: Followed the suggestion but then clippy told me to change it a bit.

**Comment by esdrubal** [comment]
> * Test when enums are specified with their fully qualified path including the library names both with and without the `()` (`lib_a::inner_lib::my_enum::MyEnum::VariantA()`)
> * Test function calls and method calls that take no arguments both with and without the `()` and both with and without their fully qualified paths including the library names

Added the tests mentioned above in most recent change. Thanks to this I found an issue where a fully qualified path method with no parameters could be called without parenthesis. This issue was fixed with a new error thrown.

File: sway-error/src/error.rs [hunk 5]
```diff
@@ -442,6 +444,8 @@ pub enum CompileError {
         expected: usize,
         received: usize,
     },
+    #[error("The function \"{method_name}\" was called without parenthesis.")]
```
Comment (emilyaherbert) [inline sway-error/src/error.rs:447]: Can you change this to:

"The function \"{method_name}\" takes no arguments, but was called without parenthesis. Try adding ()."

Thanks!

**Review by emilyaherbert — COMMENTED** [review]
Great fixes! Just one slight rewording to the new error message, then I'm happy to approve 😄

**Review by emilyaherbert — DISMISSED** [review]

File: sway-error/src/error.rs (hunk 5 above)
Comment (esdrubal) [reply sway-error/src/error.rs]: Changed to "The function "{method_name}" was called without parenthesis. Try adding ()."
Removed the "takes no arguments" as the method may or may not take arguments. This error is thrown when the method is called with no arguments but it may still take arguments.

File: sway-error/src/error.rs [hunk 6]
```diff
@@ -416,6 +416,8 @@ pub enum CompileError {
     MoreThanOneEnumInstantiator { span: Span, ty: String },
     #[error("This enum variant represents the unit type, so it should not be instantiated with any value.")]
     UnnecessaryEnumInstantiator { span: Span },
+    #[error("The enum variant `{ty}` is of type `unit`, so its constructor does not take arguments or parenthesis. Try removing the ().")]
```
Comment (mohammadfawaz) [inline sway-error/src/error.rs:419]:
```suggestion
    #[error("The enum variant `{ty}` is of type `unit`, so its constructor does not take arguments or parentheses. Try removing the ().")]
```

File: sway-error/src/error.rs [hunk 7]
```diff
@@ -442,6 +444,8 @@ pub enum CompileError {
         expected: usize,
         received: usize,
     },
+    #[error("The function \"{method_name}\" was called without parenthesis. Try adding ().")]
+    MissingParenthesisForFunction { span: Span, method_name: Ident },
```
Comment (mohammadfawaz) [inline sway-error/src/error.rs:448]:
```suggestion
    #[error("The function \"{method_name}\" was called without parentheses. Try adding ().")]
    MissingParenthesesForFunction { span: Span, method_name: Ident },
```

File: sway-error/src/error.rs [hunk 8]
```diff
@@ -813,10 +817,12 @@ impl Spanned for CompileError {
             ImportMustBeLibrary { span, .. } => span.clone(),
             MoreThanOneEnumInstantiator { span, .. } => span.clone(),
             UnnecessaryEnumInstantiator { span, .. } => span.clone(),
+            UnitVariantWithParenthesesEnumInstantiator { span, .. } => span.clone(),
             TraitNotFound { span, .. } => span.clone(),
             InvalidExpressionOnLhs { span, .. } => span.clone(),
             TooManyArgumentsForFunction { span, .. } => span.clone(),
             TooFewArgumentsForFunction { span, .. } => span.clone(),
+            MissingParenthesisForFunction { span, .. } => span.clone(),
```
Comment (mohammadfawaz) [inline sway-error/src/error.rs:825]:
```suggestion
            MissingParenthesesForFunction { span, .. } => span.clone(),
```

File: sway-core/src/semantic_analysis/ast_node/expression/typed_expression/function_application.rs [hunk 9]
```diff
@@ -22,6 +22,15 @@ pub(crate) fn instantiate_function_application(
     let decl_engine = ctx.decl_engine;
     let engines = ctx.engines();
 
+    if arguments.is_none() {
+        errors.push(CompileError::MissingParenthesisForFunction {
```
Comment (mohammadfawaz) [inline sway-core/src/semantic_analysis/ast_node/expression/typed_expression/function_application.rs:26]:
```suggestion
        errors.push(CompileError::MissingParenthesesForFunction {
```

**Review by mohammadfawaz — COMMENTED** [review]
Looks great! Just some nits and then I'll approve.

**Review by mohammadfawaz — APPROVED** [review]

**Review by JoshuaBatty — APPROVED** [review]
